### PR TITLE
Rework locks

### PR DESCRIPTION
**Pull request by bugadani**
- Move locks to a new module
- Deduplicate implementation
- Reimplement critical_section in terms of Lock
- Make Lock reentrant by default

This change should prevent multi-core apps spinning in interrupt-free context if the other core is holding the global lock.

It's hardly a stress test, but I've ran the multicore examples on an ESP32 and they worked as expected.

### CODE REVIEW

File: esp-hal/src/lock.rs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+            if #[cfg(riscv)] {
+                let mut mstatus = 0u32;
+                core::arch::asm!("csrrci {0}, mstatus, 8", inout(reg) mstatus);
+                ((mstatus & 0b1000) != 0) as critical_section::RawRestoreState
+            } else if #[cfg(xtensa)] {
+                let token: critical_section::RawRestoreState;
+                core::arch::asm!("rsil {0}, 5", out(reg) token);
+                token
+            } else {
+                panic!()
```
Comment (Dominaezzz) [inline esp-hal/src/lock.rs:13]:
```suggestion
                compile_error!("");
```
?

**Comment by bugadani** [comment]
Fun fact: this PR came 2 days before #797 turned one year old :D

**Comment by MabezDev** [comment]
> Fun fact: this PR came 2 days before https://github.com/esp-rs/esp-hal/pull/797 turned one year old :D

I expect you to mark this occasion in your calendar for next year too :D

File: esp-hal/src/timer/timg.rs [hunk 2]
```diff
@@ -76,19 +76,18 @@ use crate::soc::constants::TIMG_DEFAULT_CLK_SRC;
...
     system::{Peripheral as PeripheralEnable, PeripheralClockControl},
     Async,
     Blocking,
     InterruptConfigurable,
-    LockState,
     Mode,
 };
 
-static INT_ENA_LOCK: LockState = LockState::new();
+static INT_ENA_LOCK: Lock = Lock::new();
```
Comment (Dominaezzz) [inline esp-hal/src/timer/timg.rs:90]: Whoops, this should've been one lock per timer group rather than one for both.

I'd do it in another PR, but since you've made other drive-by changes.... :smile: .. might as well do it here? :sweat_smile:

**Review by bjoernQ — APPROVED** [review]
Nice!

**Review by MabezDev — COMMENTED** [review]
LGTM, I'm glad we're reusing code for both the stand alone lock and CS now.

It would be nice to have a test or two, but we can leave that for now. One final thing to consider before merging, what about naming the lock module `sync` instead?

**Comment by bugadani** [comment]
> One final thing to consider before merging, what about naming the lock module sync instead?

It's private, we can deal with this later, when we're fixing the fallout from this 🤣

**Review by MabezDev — APPROVED** [review]
